tests/kernel/interrupt/src/direct_isr.c

```diff
@@ -0,0 +1,139 @@
+#define TEST_DIRECT_IRQ_LINE_2	4
+
+/* the vector got would be different when enable code coverage */
+#if defined(CONFIG_COVERAGE)
+#define TRIGGER_DIRECT_IRQ_LINE_1	33
+#define TRIGGER_DIRECT_IRQ_LINE_2	34
+#else
+#define TRIGGER_DIRECT_IRQ_LINE_1	34
+#define TRIGGER_DIRECT_IRQ_LINE_2	33
+#endif
+
+#define TEST_DIRECT_IRQ_PRIO	0
+
+#elif defined(CONFIG_ARCH_POSIX)
+#define TEST_DIRECT_IRQ_LINE_1	5
+#define TEST_DIRECT_IRQ_LINE_2	6
+
+#define TRIGGER_DIRECT_IRQ_LINE_1	TEST_DIRECT_IRQ_LINE_1
+#define TRIGGER_DIRECT_IRQ_LINE_2	TEST_DIRECT_IRQ_LINE_2
+
+#define TEST_DIRECT_IRQ_PRIO	1
+
+#endif
+
+volatile uint32_t direct_int_executed[2];
+
+ISR_DIRECT_DECLARE(direct_isr1)
+{
+	direct_int_executed[0]++;
+
+	return 0;
+}
+
+ISR_DIRECT_DECLARE(direct_isr2)
+{
+	direct_int_executed[1]++;
+
+	return 1;
+}
+
+/**
+ * @brief Test direct interrupt function
+ *
+ * @details validate direct interrupt works as expected.
+ * - Register two direct interrupt at build time.
+ * - Trigger interrupt and check if isr handler has executed or not.
+ * - Also check irq_enable and irq_disable works.
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @see IRQ_DIRECT_CONNECT(), ISR_DIRECT_DECLARE()
+ */
+void test_direct_interrupt(void)
+{
+	int trig_vec1, trig_vec2;
+
+	IRQ_DIRECT_CONNECT(TEST_DIRECT_IRQ_LINE_1, TEST_DIRECT_IRQ_PRIO, direct_isr1, 0);
+	IRQ_DIRECT_CONNECT(TEST_DIRECT_IRQ_LINE_2, TEST_DIRECT_IRQ_PRIO, direct_isr2, 0);
+
+#if defined(CONFIG_X86)
+	trig_vec1 = Z_IRQ_TO_INTERRUPT_VECTOR(TEST_DIRECT_IRQ_LINE_1);
+	trig_vec2 = Z_IRQ_TO_INTERRUPT_VECTOR(TEST_DIRECT_IRQ_LINE_2);
+#elif defined(CONFIG_ARCH_POSIX)
+	trig_vec1 = TRIGGER_DIRECT_IRQ_LINE_1;
+	trig_vec2 = TRIGGER_DIRECT_IRQ_LINE_2;
+#endif
+	TC_PRINT("irq(%d)=vector(%d)\n", TEST_DIRECT_IRQ_LINE_1, trig_vec1);
+	TC_PRINT("irq(%d)=vector(%d)\n", TEST_DIRECT_IRQ_LINE_2, trig_vec2);
+
+	/* verify the target triggering vector is correct */
+	zassert_equal(trig_vec1, TRIGGER_DIRECT_IRQ_LINE_1,
+			"vector %d mismatch we specified to trigger %d",
+			trig_vec1, TRIGGER_DIRECT_IRQ_LINE_1);
+
+	zassert_equal(trig_vec2, TRIGGER_DIRECT_IRQ_LINE_2,
+			"vector %d mismatch we specified to trigger %d",
+			trig_vec2, TRIGGER_DIRECT_IRQ_LINE_2);
+
+	irq_enable(TEST_DIRECT_IRQ_LINE_1);
+	irq_enable(TEST_DIRECT_IRQ_LINE_2);
+
+	zassert_true(direct_int_executed[0] == 0 &&
+			direct_int_executed[1] == 0,
+			"Both ISR should not execute");
+
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_1);
+
+	zassert_true(direct_int_executed[0] == 1 &&
+			direct_int_executed[1] == 0,
+			"ISR1 should execute");
+
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_2);
+
+	zassert_true(direct_int_executed[0] == 1 &&
+			direct_int_executed[1] == 1,
+			"Both ISR should execute");
+
+	irq_disable(TEST_DIRECT_IRQ_LINE_1);
+	irq_disable(TEST_DIRECT_IRQ_LINE_2);
+
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_1);
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_2);
+
+	/*
+	 * irq_enable()/irq_disable() does not work here,
+	 * see #33901.
+	 */
+#if !defined(CONFIG_X86)
```

Comment:
I'm not sure if we can allow to merge the test with such workaround. I had a bit of look at issue #33901. Seems like no further discussions about it.
@enjiamai you have to ask  @andyross about further solution of the #33901

Reply:
Yes, @maksimmasalski, let me ask him tonight.  And you talk about a point. I really want to make sure that can the testcases be merged even the CI does not pass due to a bug? Thanks so much for your comments.

Reply:
Yes, ask him please. Test case looks decent, but that workaround is something that necessary to solve.